### PR TITLE
dxList - search box implementation

### PR DESCRIPTION
**Pull request by Alyar666**

### CODE REVIEW

File: js/ui/list/ui.list.edit.js [hunk 1]
```diff
@@ -169,7 +172,47 @@ var ListEdit = ListBase.inherit({
...
             * @default false
             */
-            allowItemReordering: false
+            allowItemReordering: false,
+
+            /**
+            * @name dxListOptions_searchMode
+            * @publicName searchMode
+            * @type string
+            * @default 'contains'
```
Comment (vconst) [inline js/ui/list/ui.list.edit.js]: Where is available values?

File: js/ui/list/ui.list.edit.js [hunk 2]
```diff
@@ -281,6 +324,14 @@ var ListEdit = ListBase.inherit({
         }
     },
 
+    _dataSourceOptions: function() {
+        return extend(this.callBase(), {
+            searchValue: this.option("searchValue"),
+            searchOperations: this.option("searchMode"),
```
Comment (vconst) [inline js/ui/list/ui.list.edit.js]: Why is searchOperations?

File: js/ui/list/ui.list.edit.js [hunk 3]
```diff
@@ -318,6 +370,69 @@ var ListEdit = ListBase.inherit({
...
+
+    _renderDimensions: function() {
+        this.callBase();
+
+        var searchEditorHeight,
+            $element = this.element(),
+            $scrollableWrapper = $element && $element.children("." + SCROLLABLE_WRAPPER_CLASS);
+
+        if($scrollableWrapper && this._$searchEditorElement) {
+            searchEditorHeight = this._$searchEditorElement.outerHeight(true);
```
Comment (vconst) [inline js/ui/list/ui.list.edit.js]: Getting height is not good. Try change to use css only.

File: js/ui/list/ui.list.edit.js [hunk 4]
```diff
@@ -318,6 +370,69 @@ var ListEdit = ListBase.inherit({
...
+
+        editorOptions.onFocusIn = function() {
+            that._toggleFocusClass(true, $element);
+            userEditorOptions.onFocusIn && userEditorOptions.onFocusIn.apply(this, arguments);
+        };
+
+        editorOptions.onFocusOut = function() {
+            that._toggleFocusClass(false, $element);
+            userEditorOptions.onFocusOut && userEditorOptions.onFocusOut.apply(this, arguments);
+        };
```
Comment (vconst) [inline js/ui/list/ui.list.edit.js]: Can we remove this code without any problems?

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by mrjohn0011 — COMMENTED** [review]
I think the dxList is the simple widget designed to show collection data. It is not a good idea to make it more complex. Search functionality should be a separate module or example but not out-of-the-box functionality

**Comment by kvet** [comment]
It seems a bit strange to embed TextBox to the List. The primary List feature is to show a list. It is used by 100% of customers. The embedded search will be used by, for example, 0.1% of customers. Why should we slow down the widget for all of the customers?

**Comment by ghost** [comment]
Guys, do you use List without searching capability?
Or DevExtreme list is just control to show 3-5 items? Am I right? There is no objections if it is...

**Comment by mrjohn0011** [comment]
@poltp, there is an [example](https://js.devexpress.com/Demos/WidgetsGallery/Demo/List/ListWithSearchBar/jQuery/Light/) demonstrating how to use the search box in the dxList. It contains less of code than changes requested by @Alyar666

**Review by mrjohn0011 — CHANGES_REQUESTED** [review]
I think this changes should not be merged because this functionality can be implemented by the user with less of code as it is shown in our example

File: js/ui/list/ui.list.edit.search.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        switch(args.name) {
+            case "searchEnabled":
+            case "searchEditorOptions":
+                this._invalidate();
+                break;
+            case "searchExpr":
+            case "searchMode":
+            case "searchValue":
+                if(!this._dataSource) {
+                    errors.log("W1009");
```
Comment (mrjohn0011) [inline js/ui/list/ui.list.edit.search.js]: dxList can work with "items" option and have no value in the "dataSource" option

**Comment by ghost** [comment]
@mrjohn0011
I do not understand your logic. What is the goal of dxList?
Looks like "to show list of items", am I right?
What number of items do you show (you did not answer me)?
Ok, you show 3 or 5 items and you are right - searching capability is too complex for this control.
I suppose, 80% customers show more than 5 items. If I'm right, searching capability must be main feature of this control.

Do you use something like car or phone?
If you buy car and seller ask you to buy steering wheel separately? 
If you buy phone and seller sells you just chip and ask you buy the body separately?

Guys, I don't know your users. Our dashboard users want to work fast. As your user I suggest use your controls without coding - you are coding for the DevExpress money, I'm also do it and I don't want to get this money from DevExpress twice:)

**Comment by kvet** [comment]
@poltp 
Our goal is to release fast robust and lightweight widgets that fit the needs of the most customers. You can easily combine them to implement search panel for dxList. It's shown in our demo. And it is simple.

An analogy. You are asking us to equip cars with armored glasses and bodywork. While you will be happy in safety, many others will not buy our car as it too slow and expensive.

Of course, dxList can be better and should evolve. But as an example, I would rather have embedded virtualization. As it will increase dxList performance for many customers that pass more than 5 items to the widget.

**Comment by dxbykov** [comment]
Guys, why can't we treat this search capability as an extension/plugin to the dxList widget? This way we'll keep the widget fast and light for those who don't need it. At the same time, we'll provide an easy search capabilities integration for those who need it. In the future, we should apply this approach to other DevExtreme widgets, so a widget feature can be enabled by dynamically added extensions not built-in statically. Keep this in mind to minimize the amount of breaking changes in the future.

File: js/ui/list/ui.list.edit.search.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            * @publicName searchEnabled
+            * @type boolean
+            * @default false
+            */
+            searchEnabled: false,
+
+            /**
+             * @name dxListOptions_searchEditorOptions
+             * @publicName searchEditorOptions
+             * @type TextBox options
```
Comment (lukyanovas) [inline js/ui/list/ui.list.edit.search.js]: change to dxTextBoxOptions

**Review by lukyanovas — CHANGES_REQUESTED** [review]

**Comment by kaatula** [comment]
@mrjohn0011  @kvet  you could miss that we should deliver not only lightweight widgets, but - in the same time - useful widgets. And that our widgets should be ready for big data, or whatever is exact quotation. 

The opinion of people who tried to **really** use List with more than a couple of dozens items is strangelly unique.

If @dxbykov idea is to provide such search **extension** out of the box, so the only difference with `not built-in statically` feature that it won't break on JS interpretation if required module is missed - I'm for it.

But if you guys decided to embed virtualization in this iteration too - that's a great news!

**Review by mrjohn0011 — DISMISSED** [review]

**Comment by kaatula** [comment]
@mrjohn0011  are referreng to this example, which demonstrates how to integrate dxList and dxTextBox and keep it with perfect common UI style?
image

**Comment by mrjohn0011** [comment]
I approved this PR if it is going to be a separate plugin in the future

**Review by lukyanovas — APPROVED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**